src/Microsoft.ML.Parquet/ParquetLoader.cs

```diff
@@ -390,9 +390,21 @@ public Cursor(ParquetLoader parent, Func<int, bool> predicate, IRandom rand)
                     Columns = _loader._columnsLoaded.Select(i => i.Name).ToArray()
                 };
 
-                int numBlocks = (int)Math.Ceiling(((decimal)parent.GetRowCount() / _readerOptions.Count));
-                int[] blockOrder = _rand == null ? Utils.GetIdentityPermutation(numBlocks) : Utils.GetRandomPermutation(rand, numBlocks);
-                _blockEnumerator = blockOrder.GetEnumerator();
+                try
+                {
+                    int numBlocks = checked((int)Math.Ceiling(((decimal)parent.GetRowCount() / _readerOptions.Count)));
```

Comment:
Elsewhere, our checks check whether something would result in `Utils.ArrayMaxLength`. Of course those checks also assume that the user is running with `gcAllowVeryLargeObjects` in the app config, which may not be the case in API usage... :P

---
In reply to: [186318365](https://github.com/dotnet/machinelearning/pull/12#discussion_r186318365) [](ancestors = 186318365)

Reply:
Similar to another comment, we wanted the {block size} and {number of blocks} to both be under 300M to fit in an int array so that it can be shuffled. I changed the default block size to be 1M, but the user could change that value, and if either exceeds the upper limit the exception will be thrown :P